### PR TITLE
winConsoleUIA: Greatly speed up move by word

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
None.

### Summary of the issue:
In Windows consoles with UIA enabled, moving the review cursor by word is extremely slow, especially at the end of long lines. Sometimes it can take more than 250 ms to find the next or previous word.
The reason for this slowness is that NVDA must manually calculate word positions within a line because the UIA implementation for consoles in Windows does not support moving by word, and within NVDA's calculation, it needs to work out the current character offset within a line, for which it currently calls IUIAutomationTextRange::move with unit character, over and over again until reaching the start of the line. This call to move could happen up to 80 times before finding the start of the line.

### Description of how this pull request fixes the issue:
Rather than calling move over and over again to find the start of the line, just fetch the text from the start of the line up to the current position, and use the number of characters. Specifically, the number of utf16 characters (using WideStringOffsetConvertor), as each character cell in the console holds one utf16 character.
 
### Testing performed:
In wsl, ran the 'ls' command in a big directory, and reviewed the lines by moving by previous and next word.
Echoed several strings to the console using the 'echo' command, including emoji characters, and moved through them with next and previous word, ensuring that  the review cursor correctly moved by word, not getting stuck or jumping too far.

### Known issues with pull request:
None.

### Change log entry:
None.

### CODE REVIEW

File: source/NVDAObjects/UIA/winConsoleUIA.py [hunk 1]
```diff
@@ -232,16 +221,17 @@ def _getWordOffsetsInThisLine(self, offset, lineInfo):
 		# not more than two alphanumeric chars in a row.
 		# Inject two alphanumeric characters at the end to fix this.
 		lineText += "xx"
+		lineTextLen = textUtils.WideStringOffsetConverter(lineText).wideStringLength
```
Comment (leonardder) [inline source/NVDAObjects/UIA/winConsoleUIA.py]: Good catch

File: source/NVDAObjects/UIA/winConsoleUIA.py [hunk 2]
```diff
@@ -204,22 +205,10 @@ def _getCurrentOffsetInThisLine(self, lineInfo):
...
-				lineInfo,
-				"startToStart"
-			) >= 0:
-				res += chars
-			else:
-				break
-		return res
+		charInfo.setEndPoint(lineInfo, "startToStart")
+		text = charInfo.text
+		offset = textUtils.WideStringOffsetConverter(text).wideStringLength - 1
```
Comment (leonardder) [inline source/NVDAObjects/UIA/winConsoleUIA.py]: At first glance, it's not entirely clear to me where the -1 is coming from here.

**Comment by michaelDCurran** [comment]
The minus 1 seemed to be necessary as charInfo ended up including the 
character at the current position as well, due to collapsing bugs.

However, I have now refactored it so that rather than charInfo being 
current position stretched back to the start of the line, it is now the 
line, cut off before the current position. Thus the minus 1 is no longer 
needed as it avoids the bug completely.

**Review by LeonarddeR — APPROVED** [review]
Thanks for this solution. While looking more readable, it also fixes another concern, namely that offset somehow could be -1, theoretically. That's no longer the case now.

**Pull request merged**